numpy/fft/_pocketfft.c

```diff
@@ -2193,14 +2193,44 @@ WARN_UNUSED_RESULT static int rfft_forward(rfft_plan plan, double c[], double fc
   }
 
 static PyObject *
-execute_complex(PyObject *a1, int is_forward, double fct)
+execute_complex(PyObject *a1, int is_forward, double fct, PyObject *out)
 {
-    PyArrayObject *data = (PyArrayObject *)PyArray_FromAny(a1,
-            PyArray_DescrFromType(NPY_CDOUBLE), 1, 0,
-            NPY_ARRAY_ENSURECOPY | NPY_ARRAY_DEFAULT |
-            NPY_ARRAY_ENSUREARRAY | NPY_ARRAY_FORCECAST,
-            NULL);
-    if (!data) return NULL;
+    PyArrayObject *data;
+    if(out == Py_None) {
+        data = (PyArrayObject *)PyArray_FromAny(a1,
+              PyArray_DescrFromType(NPY_CDOUBLE), 1, 0,
+              NPY_ARRAY_ENSURECOPY | NPY_ARRAY_DEFAULT |
+              NPY_ARRAY_ENSUREARRAY | NPY_ARRAY_FORCECAST,
+              NULL);
+        if (!data) return NULL;
+    }
+    else if (!PyArray_Check(out)) {
+      PyErr_SetString(PyExc_TypeError, "'out' must be an array");
+      return NULL;
+    }
+    else {
+      data = (PyArrayObject*)out;
+      int ndim = PyArray_NDIM((PyArrayObject*)a1);
+      if (PyArray_NDIM(data) != ndim) {
+          PyErr_SetString(PyExc_ValueError,
+                          "Output array has wrong dimensionality");
+          return NULL;
+      }
+      else if (!PyArray_CompareLists(PyArray_SHAPE(data), PyArray_SHAPE((PyArrayObject*)a1), ndim)) {
```

Comment:
I'm pretty sure you also need to ensure the array is C-ordered (something which the copy does for the regular case).

There may be an internal helper routine for this - let me ping @seberg (who would at least know who to ask!).

Note that it may be easier to do these tests in the python routine and here assume it is OK.

Reply:
p.s. Best to add some actual tests with wrongly constructed out, to see what fails!